### PR TITLE
Harden privileged startup against symlink-following

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
- Validate the run directory before it is used: it must be a real directory (never a symlink) with a trusted parent, and under a world-writable parent such as /tmp it must be owned by root, by us, or by the uid we drop to.
- Repair directory ownership through descriptors (fchown, fchownat(AT_SYMLINK_NOFOLLOW)) instead of paths, so a symlinked entry can no longer redirect a chown() that runs as root.

---
## Summary by cubic
Hardened startup to prevent symlink-following when creating or chowning the run directory and child files. This blocks local users from redirecting privileged chown or socket binds via symlinks.

- **Bug Fixes**
  - Validate the run directory before use: reject symlinked or non-directory leaves and unsafe parents; allow sticky parents (like `/tmp`) only if the dir is owned by root/us/the target drop-uid.
  - Open and chown via descriptors, not paths: use `os_open_dir_privileged()` with `O_NOFOLLOW` when needed, and `fchown`/`fchownat(AT_SYMLINK_NOFOLLOW)` to avoid following symlinks.
  - Trim trailing slashes and dot components with `os_dir_path_trim()` so checks apply to the entry itself and cannot be bypassed.
  - Declare the target drop-uid early with `os_run_dir_set_target_uid()` so the `/tmp` fallback created on first run is still recognized as ours on restart.
  - Repair ownership of required directories via safe descriptor-based traversal; pidfile chown now happens via its open fd.
  - Spawn server now always uses the validated `os_run_dir(true)`; removed ad-hoc `/tmp` fallback.
  - Added and refactored `run-dir-safety-test` to cover symlink attacks, normalization edge cases (`.`/`..`, mixed dot/slash), sticky-parent rules, and restart ownership; added direct tests for `os_run_dir_is_safe()`.

<sup>Written for commit 107279c4ddd85e65fd9411e0ff6d155b7669adaf. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/23334?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/libnetdata/os/privileged_dir.c [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    int flags = O_RDONLY | O_DIRECTORY | O_CLOEXEC;
+
+    // Following a symlink at the final component is only safe when we are the
+    // only account that could have put it there. So /var/cache/netdata may point
+    // to another filesystem (its parent /var/cache is root-only), while
+    // /var/lib/netdata/cloud.d may not (its parent is writable by netdata).
+    if(os_dir_parent_trust(dir) != OS_DIR_PARENT_EXCLUSIVE)
+        flags |= O_NOFOLLOW;
+
+    return open(dir, flags);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/privileged_dir.c]: False positive. The tainted value here is the operator's own configuration (NETDATA_RUN_DIR, or a directory from netdata.conf) - the same trust level as the config file itself, not attacker-controlled input.

Line 121 is the mitigation, not the exposure: os_open_dir_privileged() trims the path so the final component cannot be resolved as a directory, classifies the containing directory's trust, and adds O_NOFOLLOW unless that directory is writable by nobody but us. The alert fires on the hardened open() that replaced the unguarded ones.

Consistent with the existing dismissals of this rule for env-var-derived paths in this repo (#3329, #3045, #3211, #3210). Dismissing as false positive.

**Comment by stelfrag** [comment]
@cubic-dev-ai review PR

File: src/libnetdata/os/run_dir.c [hunk 2]
```diff
@@ -22,30 +44,114 @@ static inline bool is_dir_accessible(const char *dir, bool rw) {
...
+
+        case OS_DIR_PARENT_STICKY:
+            // The /tmp fallback. Anyone can create entries here, but sticky
+            // means only the owner of an entry can rename or delete it, so
+            // requiring the directory to be ours also closes the window between
+            // this check and the privileged use that follows it (the temporary
+            // spawn server binds its socket here long before become_user()).
+            // "Ours" has to include the uid we will drop to, because the first
+            // run creates this directory as root and then chowns it.
+            if (run_dir_owner_is_ours(st.st_uid))
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/run_dir.c:183]: False positive as written: the leaf's own mode is already checked. `os_run_dir_is_safe()` refuses `S_IWOTH` before it ever looks at the parent (run_dir.c:90), with an error log, and the test asserts it for both 0777 and 01777 leaves (`directory writable by everyone`, `directory writable by everyone, even sticky`).

The suggested edit also adds `S_IWGRP`, which we cannot take: the shipped systemd unit creates /run/netdata with `RuntimeDirectoryMode=0775` and re-applies it on every start, so refusing group-write would leave the default install with no run directory at all. That is deliberate and covered by the `directory writable by its group (systemd 0775 shape)` case.

[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/run_dir.c:183]: That is exactly what this branch does - the pathname is never what the privileged operation acts on.

Accepting a target-owned leaf under a sticky parent is required (the first run creates /tmp/netdata as root and chowns it, so on restart it is owned by the drop-to uid), and the code says so at run_dir.c:106-118 including this residual. The protection is the descriptor:

- `os_open_dir_privileged()` opens the parent once, judges it through `fstat()` on that descriptor, and looks the leaf up with `openat()` - adding `O_NOFOLLOW` whenever the parent is not EXCLUSIVE, which is precisely the sticky /tmp case. A symlink swapped in after validation gets ELOOP, it is not followed.
- Everything privileged then goes through that fd: `fchown()` for the ownership fix (daemon.c:62), `unlinkat()`/`fchmodat(AT_SYMLINK_NOFOLLOW)` for the socket, and `bind()` through /proc/self/fd/<dir_fd> when the parent is not exclusive (spawn_server_nofork.c:1455+).

Both guarantees are asserted, not assumed: `run-dir-safety-test` has `symlink in a world-writable parent (refused)`, `symlink refused, its target left alone` for fchmodat, and `run dir swapped after open, before bind`, which renames the directory away and plants a symlink in its place before binding.

The residual after that is swapping in another real directory owned by the same uid - which is the account we are about to become and which already owned the directory we validated, so it gains nothing.

File: src/libnetdata/spawn_server/spawn_server_nofork.c [hunk 3]
```diff
@@ -1389,34 +1391,17 @@ SPAWN_SERVER* spawn_server_create(SPAWN_SERVER_OPTIONS options, const char *name
...
-            // it does not exist
-            nd_log(NDLS_COLLECTORS, NDLP_ERR, "Runtime directory '%s' does not exist, falling back to '/tmp'", runtime_directory);
-            runtime_directory = NULL;
-        }
+    // os_run_dir() already validates the candidate (including NETDATA_RUN_DIR) and
+    // refuses a symlinked run dir. Never fall back to /tmp directly: the socket
+    // name is predictable, so a world-writable directory would let any local user
+    // pre-plant it.
+    const char *runtime_directory = os_run_dir(true);
+    if(!runtime_directory || !*runtime_directory) {
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/spawn_server/spawn_server_nofork.c:1587]: Valid - fixed in f3f1b15, at the source rather than in the caller.

Root cause was in `os_run_dir()`: the cache had one slot and no record of what it had validated, so the first caller's answer was returned to everyone. A `rw=false` resolution creates nothing and only checks `R_OK`, so it can settle on a directory we cannot write into (read-only /run bind mount, root-owned /run/netdata under a non-root agent) or on the /tmp fallback while /run/netdata was never created - and it also skips the `nd_setenv()` that tells children where the run dir is.

The cache now records whether it satisfied a writable request; the first `rw=true` caller re-runs the detection, and gets NULL if no writable directory exists rather than a path it cannot use. Fixing it there covers every writable consumer (spawn server, `change_dir_ownership()`, the netipc servers) instead of only this call site. Revalidating in the caller was the other option offered, but it would leave the wrong answer cached for the next one.

Regression case added to `run-dir-safety-test` (`writable request after a read-only one`); verified it fails against the pre-fix file and passes after.

Note this was latent in-tree: the daemon resolves `rw=true` first (main.c:1034, before the spawn server), so no current ordering hit it.

File: src/libnetdata/os/privileged_dir.c [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    if(!os_dir_path_trim(path, dir, sizeof(dir)))
+        // errno is already EINVAL (nothing left to name) or ENAMETOOLONG
+        return -1;
+
+    int flags = O_RDONLY | O_DIRECTORY | O_CLOEXEC;
+
+    char parent[FILENAME_MAX + 1];
+    const char *leaf;
+    if(!os_dir_split_parent(dir, parent, sizeof(parent), &leaf))
+        return open(dir, flags); // dir is "/": there is no final component to protect
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/privileged_dir.c:146]: Dismissed as a false positive in code scanning.

The taint source is `getenv("NETDATA_RUN_DIR")`, which is operator configuration on the same side of the trust boundary as the process: anything able to set the agent's environment can already execute as that account. It is a documented knob for hardened containers with a read-only /run.

The value is not used raw either. `os_run_dir_is_safe()` validates it before this function ever sees it (trimmed so a trailing slash cannot hide the final component, `lstat()` so a symlinked leaf is refused, no world-writable leaf, and an ownership/parent-trust check), and `os_open_dir_privileged()` then opens the parent once, judges it with `fstat()` on that descriptor and looks the leaf up with `openat()` - adding `O_NOFOLLOW` whenever the parent is not exclusively ours. So the directory that was judged is the directory that gets used, and every privileged operation afterwards goes through that descriptor (`fchown`, `unlinkat`, `fchmodat(AT_SYMLINK_NOFOLLOW)`, `bind` via /proc/self/fd).

File: src/libnetdata/os/privileged_dir.c [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    // Judge and use the same directory. Deciding on stat(parent) and then
+    // open()ing the whole path resolves the intermediate components a second
+    // time, so the directory we trusted need not be the one the final component
+    // is then looked up in: anyone able to retarget or rename a component above
+    // the parent can make the two disagree. Holding one descriptor and looking
+    // the entry up through it removes that gap.
+    int parent_fd = open(parent, O_RDONLY | O_DIRECTORY | O_CLOEXEC);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/privileged_dir.c:154]: Dismissed as a false positive in code scanning.

The taint source is `getenv("NETDATA_RUN_DIR")`, which is operator configuration on the same side of the trust boundary as the process: anything able to set the agent's environment can already execute as that account. It is a documented knob for hardened containers with a read-only /run.

The value is not used raw either. `os_run_dir_is_safe()` validates it before this function ever sees it (trimmed so a trailing slash cannot hide the final component, `lstat()` so a symlinked leaf is refused, no world-writable leaf, and an ownership/parent-trust check), and `os_open_dir_privileged()` then opens the parent once, judges it with `fstat()` on that descriptor and looks the leaf up with `openat()` - adding `O_NOFOLLOW` whenever the parent is not exclusively ours. So the directory that was judged is the directory that gets used, and every privileged operation afterwards goes through that descriptor (`fchown`, `unlinkat`, `fchmodat(AT_SYMLINK_NOFOLLOW)`, `bind` via /proc/self/fd).

File: src/libnetdata/spawn_server/spawn_server_nofork.c [hunk 6]
```diff
@@ -1312,12 +1314,139 @@ void spawn_server_destroy(SPAWN_SERVER *server) {
...
+// fchmodat(AT_SYMLINK_NOFOLLOW), returns ENOTSUP on glibc < 2.32 - the O_PATH
+// emulation landed there, and the kernel only grew fchmodat2() in 6.6 - which is
+// still CentOS 7, Amazon Linux 2, every RHEL 8 derivative and Debian 11.
+// umask is process-wide, so the change is serialized and held only across bind().
+// netipc's bind_owner_only_socket() does the same for its own sockets and cannot
+// be shared: that subtree deliberately carries no libnetdata dependency.
+static SPINLOCK spawn_server_umask_spinlock = SPINLOCK_INITIALIZER;
+
+static int spawn_server_bind_with_mode(int sock, const struct sockaddr_un *addr) {
+    spinlock_lock(&spawn_server_umask_spinlock);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/spawn_server/spawn_server_nofork.c:1360]: Correct that the spinlock cannot isolate the process - and that is not what it is for. It only stops this helper from nesting with itself, so the mask it saves is always the process's real one rather than another instance's temporary 0007. No lock can protect an unrelated thread's `open(O_CREAT)`, including the process-wide one suggested here, because unrelated creators would not take it.

What makes the window safe is that no concurrent creator exists at any point where the mask actually changes:

- `become_daemon()` sets `umask(0007)` permanently (daemon.c:522, from main.c:1190). For the "plugins" spawn server (main.c:1199), every lazy `spawn_popen` creation after it, and every plugin that inherits the mask, `umask(S_IRWXO)` is a no-op - same value in, same value out.
- The one creation before that point, the "init" server at main.c:1059, runs while the daemon is still single-threaded. Every thread-creating call site in `netdata_main()` is at main.c:1150 or later (sqlite, ml_init, static threads at 1294/1335, sentry at 1205, dbengine tier init via rrd_init at 1260, commands_init at 1257). `daemon_status_file_check_crash()` at 1055 posts synchronously on the main thread by design.
- The standalone users - cgroup-network.c:842, local_listeners.c:265, network-viewer.c:7334 - all bind before they create any thread (network-viewer's `functions_evloop_init()` is at 7412, after the bind at 7334).

The one real hazard nearby is nesting with netipc's `bind_owner_only_socket()`, which keeps its own lock and mask: two independent save/restore pairs on one process-wide value would restore each other's temporary. That subtree carries no libnetdata dependency by design, so the locks cannot be merged. It does not happen today - cgroups.plugin is the only process with both, and it binds its spawn server from `cgroups_detect_systemd()`/`cgroups_try_detect_version()` while reading its configuration, before the CGDISCOVER thread that calls `cgroup_netipc_init()` is created (sys_fs_cgroup.c:1403).

No code change, but the comment above the function said "umask is process-wide, so the change is serialized", which reads as though the lock provided isolation. Rewritten to state what the lock does and does not do, and to record the three reasons above plus the netipc constraint, so the next reader does not have to re-derive it.

File: src/libnetdata/os/run_dir.c [hunk 7]
```diff
@@ -22,30 +51,166 @@ static inline bool is_dir_accessible(const char *dir, bool rw) {
...
+            // check.
+            //
+            // A read-only caller is exempt: it creates nothing and chowns nothing,
+            // and it cannot answer the question anyway - netdatacli run by root
+            // does not know which uid the agent runs as, so requiring the
+            // directory to be "ours" only made it unable to find an agent that
+            // uses this fallback. Everything else above still applies to it, so it
+            // cannot be pointed elsewhere by a planted symlink, and the socket's
+            // own mode is what decides whether it may actually talk to the agent.
+            if (!rw || run_dir_owner_is_ours(st.st_uid))
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/run_dir.c]: Intentional, and the reasoning is recorded in the file above the check - but you are right that it leaves a question open, so I have tracked the part that actually answers it: #23396.

Why the exemption exists: applying the ownership rule to read-only callers broke ordinary non-root installs. A run directory under a parent owned by the netdata user (a source install, `NETDATA_RUN_DIR=~netdata/run`) is UNSAFE to a client running as root, so `netdatacli` could not find an agent that was running perfectly well.

Why the suggested variant does not work: requiring ownership when the client has euid 0 re-breaks exactly that case. The agent's `/tmp/netdata` is owned by the uid it drops to, and only the daemon knows which uid that is - it declares it with `os_run_dir_set_target_uid()`. A root `netdatacli` has no way to know it, so it would refuse the agent's own directory.

On reachability: the `/tmp` fallback is only reached when neither `/run/netdata` nor `/var/run/netdata` exists, so in practice the agent is not running or the install genuinely lives in `/tmp`. Plugins are pinned regardless - the agent exports `NETDATA_RUN_DIR` on its writable resolution and a child reads that first. And an attacker cannot take over an existing `/tmp/netdata`: `/tmp` is sticky and the directory is not theirs, so they must win the race before the agent's first start - at which point the writable check refuses the directory and the agent does not start.

Where you are right: the "socket's own mode decides" line does not settle trust, because whoever owns a planted directory also owns the socket in it and picks its mode. That is a peer-identity question, not a directory question, and the answer belongs after `connect()`: `SO_PEERCRED` / `getpeereid()`, refusing a peer that is neither root nor the agent's uid. Tightening the directory rule cannot substitute for it and costs real installs. #23396 tracks it for `netdatacli` and the netipc client handshake, which also sends its auth token in the HELLO.

File: src/libnetdata/os/tests/run-dir-safety-test.c [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+    snprintfz(p, sizeof(p), "%s/afile", exclusive);
+    check_mode("a regular file, not a directory", p, false, false);
+
+    // Accepted for a client, refused for the agent: the same directory, judged by
+    // what the caller is about to do with it. A non-root install legitimately puts
+    // its run directory under a parent the agent's own account owns, and a client
+    // that only connects must still be able to find it.
+    snprintfz(p, sizeof(p), "%s/ours", open_parent);
+    check_mode("world-writable parent without the sticky bit", p, false, true);
+    check_mode("world-writable parent without the sticky bit", p, true, false);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/tests/run-dir-safety-test.c]: Valid - fixed in 46895a9.

The collision was wider than the two pairs. `symlink to /etc`, `a regular file, not a directory`, `world-writable parent without the sticky bit` and `directory writable by everyone` each also appeared in the earlier agent-only section, so a FAILED line was ambiguous across the whole file, not just within a pair.

Every `check_mode()` case now names the caller it is judged as - `... (client)` for the read-only rule, `... (agent)` for the writable one - which is also where the two opposite outcomes come from, so the label explains the result rather than contradicting the line above it. The `check()` cases keep their labels: that helper is agent-only by construction. Swept the file afterwards; no duplicate check label remains.

File: src/libnetdata/os/run_dir.h [hunk 9]
```diff
@@ -9,9 +9,52 @@
...
+ * could have planted or could replace. Trailing slashes are trimmed internally,
+ * so a caller cannot accidentally disable the symlink check by passing one.
+ *
+ * rw is the level of validation, not just an access mode. Both levels refuse a
+ * symlinked final component, a non-directory, and anything we cannot access - no
+ * caller may be redirected by a symlink planted at this name.
+ *
+ * With rw the directory is additionally about to be created, chowned and bound to
+ * while still root, so it must also be one no other account can replace: not
+ * world-writable, under a parent that is exclusively ours, or - in a sticky parent
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/os/run_dir.h]: Valid - fixed in 46895a9. It read as three alternatives when world-writable is a mandatory check applied before the parent-trust switch. Now spelled out as two conditions that must both hold:

    Both of the following must hold:
      - it is not world-writable, whatever its parent looks like; and
      - its parent is exclusively ours, or - in a sticky parent such as the /tmp
        fallback - the directory is owned by us or by the declared "run as user".

which puts the or-branch inside the second condition, where it belongs.

File: src/daemon/config/README.md [hunk 10]
```diff
@@ -151,6 +151,23 @@ The multiplication of all the **enabled** tiers `dbengine tier N update every it
...
 |      registry       |              `/opt/netdata/var/lib/netdata/registry`               | Contains the [registry](/src/registry/README.md) database and GUID that uniquely identifies each Netdata Agent                                                                     |
 
+**Moving a directory to another filesystem.** Either set the option above to the new
+path, or bind-mount the new location onto the old path. Both are supported.
+
+**Do not use a symbolic link.** At startup, while it still runs as `root`, the Agent
+gives these directories to the `netdata` user so that it can keep writing to them
+after it drops privileges - and it does that without following symbolic links. If you
+replace one of these directories, or a directory inside one of them, with a symbolic
+link, the Agent changes the ownership of the link itself and leaves the files behind
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/config/README.md]: Valid - fixed in 46895a9. The blanket prohibition was wrong for the common case, exactly as described: `os_open_dir_privileged()` follows a final-component symlink when the directory holding it is exclusively ours, and `/var/cache`, `/var/lib` and `/var/log` are root-only, so those links are followed and the target is what gets chowned.

Rewritten to state the per-parent rule instead:

- a link that replaces one of the configured paths is followed, and the directory it points to is the one that changes ownership - `/var/cache/netdata` may be a link to another filesystem, because only root can create entries in `/var/cache`;
- a link deeper inside them - `/var/lib/netdata/cloud.d`, `/var/lib/netdata/registry`, anything inside `/var/cache/netdata` - is not followed, because the netdata user can write those directories and could therefore have planted it.

Also dropped "It reports every symbolic link it skips this way": accurate for the configured paths, which log when they cannot be opened, but not for a symlinked subdirectory found during the recursive walk, where the `openat(O_NOFOLLOW)` simply skips it (daemon.c:37). Better to not claim a report the operator will not find in the log.

The run-directory paragraph below it is unchanged and still correct - a symlink there is refused outright, whatever the parent looks like.
